tests/src/main/java/net/neoforged/neoforge/debug/transfer/VanillaHandlersTests.java

```diff
@@ -0,0 +1,566 @@
+        var hopperEntity = helper.getBlockEntity(pos, HopperBlockEntity.class);
+        var hopper = VanillaContainerWrapper.of(hopperEntity);
+
+        // Insertion into empty hopper -> cooldown
+        try (var transaction = Transaction.open(null)) {
+            hopper.insert(RESOURCE, 10, transaction);
+            transaction.commit();
+        }
+        helper.assertValueEqual(HopperBlockEntity.MOVE_ITEM_SPEED, getHopperCooldown(hopperEntity), "hopper cooldown");
+        hopperEntity.setCooldown(0);
+
+        // Second insertion into hopper -> no cooldown
+        try (var transaction = Transaction.open(null)) {
+            hopper.insert(RESOURCE, 10, transaction);
+            transaction.commit();
+        }
+        helper.assertValueEqual(0, getHopperCooldown(hopperEntity), "hopper cooldown");
+
+        hopperEntity.clearContent();
+        hopperEntity.setItem(4, RESOURCE.toStack());
+
+        // Insertion into non-empty (with an item at a different index) hopper -> no cooldown
+        try (var transaction = Transaction.open(null)) {
+            hopper.insert(RESOURCE, 10, transaction);
+            transaction.commit();
+        }
+        helper.assertValueEqual(0, getHopperCooldown(hopperEntity), "hopper cooldown");
+
+        // Extraction -> no cooldown
+        try (var transaction = Transaction.open(null)) {
+            hopper.extract(RESOURCE, 15, transaction);
+            transaction.commit();
+        }
+        helper.assertContainerEmpty(pos);
+        helper.assertValueEqual(0, getHopperCooldown(hopperEntity), "hopper cooldown");
+
+        // Simulated insertion into empty hopper -> no cooldown
+        try (var transaction = Transaction.open(null)) {
+            hopper.insert(RESOURCE, 10, transaction);
+        }
+        helper.assertValueEqual(0, getHopperCooldown(hopperEntity), "hopper cooldown");
+
+        // Insertion into empty hopper + extract in the same transaction -> cooldown
+        try (var transaction = Transaction.open(null)) {
+            hopper.insert(RESOURCE, 10, transaction);
+            helper.assertContainerContains(pos, RESOURCE.getItem());
+            hopper.extract(RESOURCE, 10, transaction);
+            transaction.commit();
+        }
+        helper.assertValueEqual(HopperBlockEntity.MOVE_ITEM_SPEED, getHopperCooldown(hopperEntity), "hopper cooldown");
+
+        helper.succeed();
+    }
+
+    private static int getHopperCooldown(HopperBlockEntity hopper) {
+        return ObfuscationReflectionHelper.getPrivateValue(HopperBlockEntity.class, hopper, "cooldownTime");
+    }
+
+    @GameTest
+    @EmptyTemplate
+    @TestHolder(description = "Ensures that furnace cook time is only reset when extraction is actually committed.")
+    public static void testFurnaceCookTime(ExtendedGameTestHelper helper) {
+        BlockPos pos = new BlockPos(0, 1, 0);
+        helper.setBlock(pos, Blocks.FURNACE.defaultBlockState());
```

Comment:
Especially with a large portion of this method being in a lambda anyway, I feel like this would make more sense having the test method take a `DynamicTest` and using `DynamicTest#registerGameTestTemplate` to specify the placed block

Reply:
I feel like this would make things a bit more complex for no good reason.